src/Features/Core/Portable/ForeachToFor/AbstractForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,346 @@
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ForeachToFor
+{
+    internal abstract class AbstractForeachToForCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        private const string get_Count = nameof(get_Count);
+        private const string get_Item = nameof(get_Item);
+
+        private const string Length = nameof(Length);
+        private const string Count = nameof(Count);
+
+        private static readonly ImmutableArray<string> s_KnownInterfaceNames =
+            new string[] { typeof(IList<>).FullName, typeof(IReadOnlyList<>).FullName, typeof(IList).FullName }.ToImmutableArray();
+
+        protected abstract SyntaxNode GetForeachStatement(SyntaxToken token);
+        protected abstract (SyntaxNode start, SyntaxNode end) GetForeachBody(SyntaxNode foreachStatement);
+        protected abstract void ConvertToForStatement(SemanticModel model, ForeachInfo info, SyntaxEditor editor, CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var foreachStatement = GetForeachStatement(token);
+            if (foreachStatement == null)
+            {
+                return;
+            }
+
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var semanticFact = document.GetLanguageService<ISemanticFactsService>();
+            var foreachInfo = GetForeachInfo(semanticFact, model, foreachStatement, cancellationToken);
+            if (foreachInfo == null)
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ForeachToForCodeAction(
+                    FeaturesResources.Convert_foreach_to_for,
+                    c => ConvertForeachToForAsync(document, foreachInfo, c)));
+        }
+
+        protected string CreateUniqueMethodName(SemanticModel model, SyntaxNode contextNode, string baseName)
+        {
+            Contract.ThrowIfNull(contextNode);
+            Contract.ThrowIfNull(baseName);
+
+            return NameGenerator.GenerateUniqueName(baseName, string.Empty,
+               name => model.LookupSymbols(contextNode.SpanStart, /*container*/null, name).Length == 0);
```

Comment:
We have some code for this already in AbstractIntroduceVariableService.  namely:

```c#
var existingSymbols = semanticModel.GetExistingSymbols(container, cancellationToken);

            var baseName = semanticFacts.GenerateNameForExpression(
                semanticModel, expression, capitalize: isConstant, cancellationToken: cancellationToken);
            var reservedNames = semanticModel.LookupSymbols(expression.SpanStart)
                                             .Select(s => s.Name)
                                             .Concat(existingSymbols.Select(s => s.Name));

            return syntaxFacts.ToIdentifierToken(
                NameGenerator.EnsureUniqueness(baseName, reservedNames, syntaxFacts.IsCaseSensitive));
```

Can we extract this out (probably to SemanticModelExtensions), so we can just share it here.

Reply:
cool. what does it do?

Reply:
It makes sure your name won't conflict with anything above, or anything below.  it also understands, for example that in "new { A = B }" that 'A' is not a name you can conflict with, even though it's a declaration of a symbol called 'A'.